test/harry/main/org/apache/cassandra/harry/dsl/PartitionVisitStateImpl.java

```diff
@@ -0,0 +1,114 @@
+/*
+  * Licensed to the Apache Software Foundation (ASF) under one
+  * or more contributor license agreements.  See the NOTICE file
+  * distributed with this work for additional information
+  * regarding copyright ownership.  The ASF licenses this file
+  * to you under the Apache License, Version 2.0 (the
+  * "License"); you may not use this file except in compliance
+  * with the License.  You may obtain a copy of the License at
+  *
+  *     http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package org.apache.cassandra.harry.dsl;
+
+import java.util.Arrays;
+import java.util.NavigableSet;
+
+import org.apache.cassandra.harry.ddl.SchemaSpec;
+
+public class PartitionVisitStateImpl implements PartitionVisitState
+{
+    final long pd;
+    final long[] possibleCds;
+    final NavigableSet<Long> visitedLts;
+    final SchemaSpec schema;
+    private final OverridingCkGenerator ckGenerator;
+
+    PartitionVisitStateImpl(long pd, long[] possibleCds, NavigableSet<Long> visitedLts, SchemaSpec schema)
+    {
+        this.pd = pd;
+        this.possibleCds = possibleCds;
+        this.visitedLts = visitedLts;
+        this.schema = schema;
+        this.ckGenerator = (OverridingCkGenerator) schema.ckGenerator;
+    }
+
+
+    /**
+     * Ensures that exactly one of the clustering keys will have given values.
+     */
+    @Override
+    public void ensureClustering(Object[] overrides)
+    {
+        long cd = getCdForOverride(overrides);
+        ckGenerator.override(cd, overrides);
+    }
+
+    @Override
+    public void overrideClusterings(Object[][] overrides)
+    {
+        assert possibleCds.length == overrides.length;
+        Arrays.sort(overrides, this::compareCds);
+        for (int i = 0; i < overrides.length; i++)
+            ckGenerator.override(possibleCds[i], overrides[i]);
+    }
+
+    @Override
+    public long pd()
+    {
+        return pd;
+    }
+
+    long getCdForOverride(Object[] ck)
```

Comment:
```suggestion
    long findCdForOverride(Object[] ck)
```
Not a big deal, but feels appropriate given the binary search?